Add terraformtype Support for github.com/hashicorp/terraform-plugin-sdk/helper/schema.StateUpgradeFunc

Version 2 of the Terraform Plugin SDK changes its type signature:

```
aws/resource_aws_db_instance.go:33:5: cannot use resourceAwsDbInstanceStateUpgradeV0 (type func(map[string]interface {}, interface {}) (map[string]interface {}, error)) as type schema.StateUpgradeFunc in field value
```

To support downstream Analyzers wishing to handle this, we can add constants and helper functions in a `helper/terraformtype/helper/schema` package similar to the others.

--- a/CHANGELOG.md
+++ b/CHANGELOG.md
@@ -9,8 +9,9 @@ ENHANCEMENTS
 
 * helper/analysisutils: `DeprecatedReceiverMethodSelectorExprRunner` now implements `SuggestedFixes`
 * helper/analysisutils: `DeprecatedWithReplacementSelectorExprRunner` now implements `SuggestedFixes`
-* helper/astutils: Add `HasFieldListLength` and `IsExprTypeBool` functions
+* helper/astutils: Add `HasFieldListLength`, `IsExprTypeBool`, `IsExprTypeMapStringInterface` functions
 * helper/terraformtype/helper/customdiff: Constants and functions for working with entire package
+* helper/terraformtype/helper/schema: Constants and functions for working with `StateUpgradeFunc` type
 * helper/terraformtype/terraform: Constants and functions for working with `ResourceProvider` type
 * passes/helper/customdiff: Passes for collecting `customdiff` function information for further analysis
 * passes/R007: Support suggested fix

--- a/helper/astutils/expr.go
+++ b/helper/astutils/expr.go
@@ -39,6 +39,13 @@ func IsExprTypeInterface(e ast.Expr) bool {
 	return ok
 }
 
+// IsExprTypeMapStringInterface returns true if the expression matches []string
+func IsExprTypeMapStringInterface(e ast.Expr) bool {
+	mapType, ok := e.(*ast.MapType)
+
+	return ok && IsExprTypeString(mapType.Key) && IsExprTypeInterface(mapType.Value)
+}
+
 // IsExprTypeString returns true if the expression matches string
 func IsExprTypeString(e ast.Expr) bool {
 	ident, ok := e.(*ast.Ident)

--- a/helper/terraformtype/helper/schema/type_stateupgradefunc.go
+++ b/helper/terraformtype/helper/schema/type_stateupgradefunc.go
@@ -1,0 +1,91 @@
+package schema
+
+import (
+	"go/ast"
+	"go/token"
+	"go/types"
+
+	"github.com/bflad/tfproviderlint/helper/astutils"
+)
+
+const (
+	TypeNameStateUpgradeFunc = `StateUpgradeFunc`
+)
+
+// IsFuncTypeStateUpgradeFunc returns true if the FuncType matches expected parameters and results types
+func IsFuncTypeStateUpgradeFunc(node ast.Node, info *types.Info) bool {
+	funcType := astutils.FuncTypeFromNode(node)
+
+	if funcType == nil {
+		return false
+	}
+
+	if !astutils.HasFieldListLength(funcType.Params, 2) {
+		return false
+	}
+
+	if !astutils.IsFieldListType(funcType.Params, 0, astutils.IsExprTypeMapStringInterface) {
+		return false
+	}
+
+	if !astutils.IsFieldListType(funcType.Params, 1, astutils.IsExprTypeInterface) {
+		return false
+	}
+
+	if !astutils.HasFieldListLength(funcType.Results, 2) {
+		return false
+	}
+
+	if !astutils.IsFieldListType(funcType.Results, 0, astutils.IsExprTypeMapStringInterface) {
+		return false
+	}
+
+	return astutils.IsFieldListType(funcType.Results, 1, astutils.IsExprTypeError)
+}
+
+// IsTypeStateUpgradeFunc returns if the type is StateUpgradeFunc from the schema package
+func IsTypeStateUpgradeFunc(t types.Type) bool {
+	switch t := t.(type) {
+	case *types.Named:
+		return IsNamedType(t, TypeNameStateUpgradeFunc)
+	case *types.Pointer:
+		return IsTypeStateUpgradeFunc(t.Elem())
+	default:
+		return false
+	}
+}
+
+// StateUpgradeFuncInfo represents all gathered StateUpgradeFunc data for easier access
+type StateUpgradeFuncInfo struct {
+	AstFuncDecl *ast.FuncDecl
+	AstFuncLit  *ast.FuncLit
+	Body        *ast.BlockStmt
+	Node        ast.Node
+	Pos         token.Pos
+	Type        *ast.FuncType
+	TypesInfo   *types.Info
+}
+
+// NewStateUpgradeFuncInfo instantiates a StateUpgradeFuncInfo
+func NewStateUpgradeFuncInfo(node ast.Node, info *types.Info) *StateUpgradeFuncInfo {
+	result := &StateUpgradeFuncInfo{
+		TypesInfo: info,
+	}
+
+	switch node := node.(type) {
+	case *ast.FuncDecl:
+		result.AstFuncDecl = node
+		result.Body = node.Body
+		result.Node = node
+		result.Pos = node.Pos()
+		result.Type = node.Type
+	case *ast.FuncLit:
+		result.AstFuncLit = node
+		result.Body = node.Body
+		result.Node = node
+		result.Pos = node.Pos()
+		result.Type = node.Type
+	}
+
+	return result
+}

--- a/helper/terraformtype/helper/schema/type_stateupgradefunc_test.go
+++ b/helper/terraformtype/helper/schema/type_stateupgradefunc_test.go
@@ -1,0 +1,177 @@
+package schema
+
+import (
+	"fmt"
+	"go/ast"
+	"go/token"
+	"go/types"
+	"testing"
+)
+
+func TestIsFuncTypeStateUpgradeFunc(t *testing.T) {
+	errorIdent := &ast.Ident{
+		Name: "error",
+	}
+	interfaceAst := &ast.InterfaceType{
+		Methods: &ast.FieldList{},
+	}
+	stringIdent := &ast.Ident{
+		Name: types.Typ[types.String].String(),
+	}
+	mapStringInterfaceAst := &ast.MapType{
+		Key:   stringIdent,
+		Value: interfaceAst,
+	}
+	typesInfo := &types.Info{
+		Types: map[ast.Expr]types.TypeAndValue{
+			errorIdent: types.TypeAndValue{
+				Type: types.NewNamed(types.NewTypeName(token.NoPos, nil, "error", nil), nil, nil),
+			},
+			interfaceAst: types.TypeAndValue{
+				Type: types.NewInterfaceType(nil, nil),
+			},
+			mapStringInterfaceAst: types.TypeAndValue{
+				Type: types.NewMap(types.Typ[types.String], types.NewInterfaceType(nil, nil)),
+			},
+			stringIdent: types.TypeAndValue{
+				Type: types.Typ[types.String],
+			},
+		},
+	}
+
+	testCases := []struct {
+		Name     string
+		Node     ast.Node
+		Info     *types.Info
+		Expected bool
+	}{
+		{
+			Name: "func(map[string]interface{}, interface{}) (map[string]interface{}, error)",
+			Node: &ast.FuncLit{
+				Type: &ast.FuncType{
+					Params: &ast.FieldList{
+						List: []*ast.Field{
+							{
+								Type: mapStringInterfaceAst,
+							},
+							{
+								Type: interfaceAst,
+							},
+						},
+					},
+					Results: &ast.FieldList{
+						List: []*ast.Field{
+							{
+								Type: mapStringInterfaceAst,
+							},
+							{
+								Type: errorIdent,
+							},
+						},
+					},
+				},
+			},
+			Info:     typesInfo,
+			Expected: true,
+		},
+		{
+			Name: "func(map[string]interface{}, interface{}) (string, error)",
+			Node: &ast.FuncLit{
+				Type: &ast.FuncType{
+					Params: &ast.FieldList{
+						List: []*ast.Field{
+							{
+								Type: mapStringInterfaceAst,
+							},
+							{
+								Type: interfaceAst,
+							},
+						},
+					},
+					Results: &ast.FieldList{
+						List: []*ast.Field{
+							{
+								Type: mapStringInterfaceAst,
+							},
+							{
+								Type: stringIdent,
+							},
+						},
+					},
+				},
+			},
+			Info:     typesInfo,
+			Expected: false,
+		},
+	}
+
+	for _, testCase := range testCases {
+		t.Run(testCase.Name, func(t *testing.T) {
+			got := IsFuncTypeStateUpgradeFunc(testCase.Node, testCase.Info)
+
+			if got != testCase.Expected {
+				t.Errorf("got %t, expected %t", got, testCase.Expected)
+			}
+		})
+	}
+}
+
+func TestIsTypeStateUpgradeFunc(t *testing.T) {
+	testCases := []struct {
+		Name      string
+		InputType types.Type
+		Expected  bool
+	}{
+		{
+			Name: fmt.Sprintf("%s.%s", PackagePath, TypeNameStateUpgradeFunc),
+			InputType: types.NewNamed(
+				types.NewTypeName(token.NoPos, types.NewPackage(PackagePath, PackageName), TypeNameStateUpgradeFunc, nil),
+				nil,
+				nil,
+			),
+			Expected: true,
+		},
+		{
+			Name: fmt.Sprintf("*%s.%s", PackagePath, TypeNameStateUpgradeFunc),
+			InputType: types.NewPointer(types.NewNamed(
+				types.NewTypeName(token.NoPos, types.NewPackage(PackagePath, PackageName), TypeNameStateUpgradeFunc, nil),
+				nil,
+				nil,
+			)),
+			Expected: true,
+		},
+		{
+			Name: fmt.Sprintf("%s.%s", PackagePath, "Not"),
+			InputType: types.NewNamed(
+				types.NewTypeName(token.NoPos, types.NewPackage(PackagePath, PackageName), "Not", nil),
+				nil,
+				nil,
+			),
+			Expected: false,
+		},
+		{
+			Name: fmt.Sprintf("%s.%s", "incorrect/path", TypeNameStateUpgradeFunc),
+			InputType: types.NewNamed(
+				types.NewTypeName(token.NoPos, types.NewPackage("incorrect/path", "path"), TypeNameStateUpgradeFunc, nil),
+				nil,
+				nil,
+			),
+			Expected: false,
+		},
+		{
+			Name:      "string",
+			InputType: types.Typ[types.String],
+			Expected:  false,
+		},
+	}
+
+	for _, testCase := range testCases {
+		t.Run(testCase.Name, func(t *testing.T) {
+			got := IsTypeStateUpgradeFunc(testCase.InputType)
+
+			if got != testCase.Expected {
+				t.Errorf("got %t, expected %t", got, testCase.Expected)
+			}
+		})
+	}
+}

--- a/passes/helper/schema/stateupgradefuncinfo/stateupgradefuncinfo.go
+++ b/passes/helper/schema/stateupgradefuncinfo/stateupgradefuncinfo.go
@@ -1,0 +1,40 @@
+package stateupgradefuncinfo
+
+import (
+	"go/ast"
+	"reflect"
+
+	"github.com/bflad/tfproviderlint/helper/terraformtype/helper/schema"
+	"golang.org/x/tools/go/analysis"
+	"golang.org/x/tools/go/analysis/passes/inspect"
+	"golang.org/x/tools/go/ast/inspector"
+)
+
+var Analyzer = &analysis.Analyzer{
+	Name: "stateupgradefuncinfo",
+	Doc:  "find github.com/hashicorp/terraform-plugin-sdk/helper/schema StateUpgradeFunc declarations for later passes",
+	Requires: []*analysis.Analyzer{
+		inspect.Analyzer,
+	},
+	Run:        run,
+	ResultType: reflect.TypeOf([]*schema.StateUpgradeFuncInfo{}),
+}
+
+func run(pass *analysis.Pass) (interface{}, error) {
+	inspect := pass.ResultOf[inspect.Analyzer].(*inspector.Inspector)
+	nodeFilter := []ast.Node{
+		(*ast.FuncDecl)(nil),
+		(*ast.FuncLit)(nil),
+	}
+	var result []*schema.StateUpgradeFuncInfo
+
+	inspect.Preorder(nodeFilter, func(n ast.Node) {
+		if !schema.IsFuncTypeStateUpgradeFunc(n, pass.TypesInfo) {
+			return
+		}
+
+		result = append(result, schema.NewStateUpgradeFuncInfo(n, pass.TypesInfo))
+	})
+
+	return result, nil
+}

--- a/passes/helper/schema/stateupgradefuncinfo/stateupgradefuncinfo_test.go
+++ b/passes/helper/schema/stateupgradefuncinfo/stateupgradefuncinfo_test.go
@@ -1,0 +1,15 @@
+package stateupgradefuncinfo
+
+import (
+	"testing"
+
+	"golang.org/x/tools/go/analysis"
+)
+
+func TestValidateAnalyzer(t *testing.T) {
+	err := analysis.Validate([]*analysis.Analyzer{Analyzer})
+
+	if err != nil {
+		t.Fatal(err)
+	}
+}

